pkg/adapterManager/manager.go

```diff
@@ -149,10 +159,27 @@ func (m *Manager) Quota(ctx context.Context, requestBag *attribute.MutableBag, r
 	return qmr, o
 }
 
+// Preprocess dispatches to the set of aspects that must run before any other
+// configured aspects.
+func (m *Manager) Preprocess(ctx context.Context, requestBag, responseBag *attribute.MutableBag) rpc.Status {
+	return m.dispatch(ctx, requestBag, responseBag, m.preprocessKindSet,
+		func(executor aspect.Executor, eval expr.Evaluator) rpc.Status {
+			ppw := executor.(aspect.PreprocessExecutor)
+			result, rpcStatus := ppw.Execute(requestBag, eval)
+			if status.IsOK(rpcStatus) {
+				if err := responseBag.Merge(result.Attrs); err != nil {
+					// TODO: better error messages that push internal details into debuginfo messages
+					rpcStatus = status.WithInternal("The results from the request preprocessing could not be merged.")
+				}
+			}
+			return rpcStatus
+		})
+}
+
 type invokeExecutorFunc func(executor aspect.Executor, evaluator expr.Evaluator) rpc.Status
 
 // dispatch resolves config and invokes the specific set of aspects necessary to service the current request
-func (m *Manager) dispatch(ctx context.Context, requestBag *attribute.MutableBag, responseBag *attribute.MutableBag,
+func (m *Manager) dispatch(ctx context.Context, requestBag, responseBag *attribute.MutableBag,
```

Comment:
I'm not convinced that we want to have special adapters that have no matching aspects and completely different configuration story. Following the story through config to managers to builders to aspects is confusing enough as is.

It seems like there may be other approaches to solving these concerns that don't involve new, distinct code paths. For instance, we could enforce, at validation time, that certain aspect kinds (`attribute-generation`, for example) cannot live anywhere but in a selector of * (must be global).